### DISCUSSION TITLE
BitOps OpsRepo generator CLI: 3 options

### DISCUSSION THREAD

**Discussion by arm4b** [bitovi/bitops#353] [Ideas]
Based on the previous discussions https://github.com/bitovi/bitops/discussions/351#discussioncomment-4099916, https://github.com/bitovi/bitops/issues/173, and https://github.com/bitovi/bitops/pull/348, we want a BitOps repo generator, - a CLI tool that will guide users through the steps helping them to create an operations repository structure.

You can find the initial quick implementation with `js` + `yo` + `npm` at https://www.bitovi.com/blog/bitops-yo

https://user-images.githubusercontent.com/1533818/201092741-0057e45e-8177-4d19-bcb2-a6950d99e279.mov

---- 
Because we need something more portable than node/js/npm for CLI and better suitable for DevOps world, long-term options for BitOps OpsRepo generator are: 

# 1) Golang CLI
### Installation
```sh
wget https://github.com/bitovi/bitops-cli/releases/download/v2.3.0/bitops-cli-$(uname -sm) \
    -O /usr/local/bin/bitops && \
  chmod +x /usr/local/bin/bitops
```
- ideally needs a dedicated repo for Golang with CI/CD pipelines
- need to build multiple binaries (OS/architecture) to upload to GH releases

### Usage
```
bitops init
```

### Pros
- Portability (just a binary to download)
- A programming language that the cloud-native community speaks
- In the DevOps world today Golang is the first choice for a CLI tool
- More likely to pique the interest of the community (vs python), - people more likely to contribute
- Elevates the team's engineering skills (it might be helpful for the future as we work deeper with K8s and its ecosystem)

### Cons
- Learning curve for a new language is present, but it's easy compared to other languages
- Maintenance and inconsistency with 2 different languages, standards, more moving pieces (`python` BitOps  + `golang` bitops-cli)

# 2) Python CLI
### Installation
```
pip install bitops-cli
```
- needs a dedicated repo with CI/CD (ideally), but could be placed in the BitOps repo (shared python)
- needs a release mechanism to deploy the package to https://pypi.org/ index
### Usage
```
bitops init
```

### Pros
- Same Python language, as the original bitops codebase
- Better maintainability

### Cons
- Needs `python` + `pip` to be installed as dependencies

# 3) BitOps Docker CLI
### Installation
No installation is needed. The generator will be bundled in the original BitOps Docker image and integrate with the project functionality as a new CLI entrypoint.

Here is an existing way how the user runs a deployment via BitOps:
```
docker run \
-e BITOPS_ENVIRONMENT="test" \
-v $(pwd):/opt/bitops_deployment
bitovi/bitops
```

Here is the proposed OpsRepo generation command: 
```
docker run \
-e BITOPS_ENVIRONMENT="test" \
-v $(pwd):/opt/bitops_deployment \
bitovi/bitops init
```
^^ find the difference.

For better user experience during the deploy stage, if the operations repo that the user shares via docker volume `-v` is empty, - suggest creating it. We can output a full docker command for copy-paste, or even go aggressive: "BitOps deployment found empty environment directory. Do you want to start creating your Operations Repository?".

### Pros
- Bundled within the original BitOps Docker image, no other tools needed
- Same BitOps repository, shared codebase, ideal maintenance and integration
- Similar Docker command used for deployment, just another CLI entrypoint parameter
- Easier transition experience in one tool: `OpsRepo generation` -> `OpsRepo development` -> `Deployment`. Users don't need to download 2 different artifacts and switch-context between the standalone CLI vs BitOps Docker image. 

### Cons
- Feels weird with Docker?
- Users have to write a longer Docker command, compared to a self-sufficient CLI binary

**Comment by PhillypHenning** [comment]
## Additional Notes
### GoLang
- Parsing logic is contained in Python, there would need to be some sort of hack applied to get access to the python parser that was created if possible, otherwise, two parsers would need to exist... 

### Docker
#### Cons
- Will likely need to deal with OS specific settings such as the Mac OS thread limit (thought that was fixed in 3.8 it still points out OS limitations)

**Reply by arm4b** [reply]
Good point about the logic that needs to be shared between the two language implementations, I missed that.

**Comment by arm4b** [comment]
Added in the https://github.com/bitovi/bitops/issues/318#issuecomment-1315229787 a few command examples how CLI can all come together as a consistent interface.

**Comment by mickmcgrath13** [comment]
for go, there are libraries like this that can help with interactive generators:
https://github.com/manifoldco/promptui

**Comment by mickmcgrath13** [comment]
Based on discussions in the Community call  (https://github.com/bitovi/bitops/discussions/370), we decided to try out Go!
